### PR TITLE
License property improvements

### LINKED ISSUE #494: Clarify default licensing value
https://github.com/Materials-Consortia/OPTIMADE/blob/b3fd0be75ee531a464d4956a781271170fd6a8f0/optimade.rst#L1263

`available_licenses` is an optional field, we should highlight what the missing value implies (e.g., database should be treated as unlicensed unless specifically listed elsewhere out-of-band).

### PR DESCRIPTION
**Pull request by merkys**
This PR addresses issues raised in #494 by:

* defining what an unknown value for `available_licenses` means;
* including the structure of the database among the things that are covered by `available_licenses`;
* introducing `available_licenses_for_entries` to specify licenses for database content and metadata only (without its structure).

Fixes #494.

### CODE REVIEW

File: optimade.rst [hunk 1]
```diff
@@ -1258,12 +1258,16 @@ The single resource object's response dictionary MUST include the following fiel
...
-    Inclusion of a license identifier in the list is a commitment of the database that the rights are in place to grant clients access to all the data and metadata according to the terms of either of these licenses (at the choice of the client).
-    If the licensing information provided via the field :field:`license` omits licensing options specified in :field:`available_licenses`, or if it otherwise contradicts them, a client MUST still be allowed to interpret the inclusion of a license in :field:`available_licenses` as a full commitment from the database that the data and metadata is available, without exceptions, under the respective licenses.
-    If the database cannot make that commitment, e.g., if only part of the data is available under a license, the corresponding license identifier MUST NOT appear in :field:`available_licenses` (but, rather, the field :field:`license` is to be used to clarify the licensing situation.)
+  - **available\_licenses**: List of `SPDX license identifiers <https://spdx.org/licenses/>` specifying a set of alternative licenses under which the client is granted access to all the content, metadata and structure of this database.
```
Comment (vaitkus) [inline optimade.rst:1261]:
```suggestion
  - **available\_licenses**: List of `SPDX license identifiers <https://spdx.org/licenses/>`__ specifying a set of alternative licenses under which the client is granted access to all the content, metadata and structure of this database.
```

File: optimade.rst [hunk 2]
```diff
@@ -1258,12 +1258,16 @@ The single resource object's response dictionary MUST include the following fiel
...
+    If the database cannot make that commitment, e.g., if only part of the content, metadata or database structure is available under a license, the corresponding license identifier MUST NOT appear in :field:`available_licenses` (but, rather, the field :field:`license` is to be used to clarify the licensing situation.)
     An empty list indicates that none of the SPDX licenses apply for the entirety of the database and that the licensing situation is clarified in human readable form in the field :field:`license`.
+    An unknown value means that the database makes no commitment to its content, metadata and/or database structure data being covered by any of the SPDX licenses.
+
+  - **available\_licenses\_for\_entries**: List of `SPDX license identifiers <https://spdx.org/licenses/>` specifying a set of alternative licenses under which the client is granted access to all the content, metadata and data extracts that neither constitute substantial parts of the database nor encompass the structure of this database.
```
Comment (vaitkus) [inline optimade.rst:1269]:
```suggestion
  - **available\_licenses\_for\_entries**: List of `SPDX license identifiers <https://spdx.org/licenses/>`__ specifying a set of alternative licenses under which the client is granted access to all the content, metadata and data extracts that neither constitute substantial parts of the database nor encompass the structure of this database.
```

File: optimade.rst [hunk 3]
```diff
@@ -1258,12 +1258,16 @@ The single resource object's response dictionary MUST include the following fiel
...
-    If the database cannot make that commitment, e.g., if only part of the data is available under a license, the corresponding license identifier MUST NOT appear in :field:`available_licenses` (but, rather, the field :field:`license` is to be used to clarify the licensing situation.)
+  - **available\_licenses**: List of `SPDX license identifiers <https://spdx.org/licenses/>` specifying a set of alternative licenses under which the client is granted access to all the content, metadata and structure of this database.
+    If the content, metadata and database structure are available under multiple alternative licenses, identifiers of these multiple licenses SHOULD be provided to let clients know under which conditions they can be used.
+    Inclusion of a license identifier in the list is a commitment of the database that the rights are in place to grant clients access to all the content, metadata and database structure according to the terms of either of these licenses (at the choice of the client).
```
Comment (vaitkus) [inline optimade.rst:1263]:
```suggestion
    Inclusion of a license identifier in the list is a commitment of the database that the rights are in place to grant clients access to all the content, metadata and database structure according to the terms of any of these licenses (at the choice of the client).
```

File: optimade.rst [hunk 4]
```diff
@@ -1258,12 +1258,16 @@ The single resource object's response dictionary MUST include the following fiel
...
+    If the database cannot make that commitment, e.g., if only part of the content, metadata or database structure is available under a license, the corresponding license identifier MUST NOT appear in :field:`available_licenses` (but, rather, the field :field:`license` is to be used to clarify the licensing situation.)
     An empty list indicates that none of the SPDX licenses apply for the entirety of the database and that the licensing situation is clarified in human readable form in the field :field:`license`.
+    An unknown value means that the database makes no commitment to its content, metadata and/or database structure data being covered by any of the SPDX licenses.
+
+  - **available\_licenses\_for\_entries**: List of `SPDX license identifiers <https://spdx.org/licenses/>` specifying a set of alternative licenses under which the client is granted access to all the content, metadata and data extracts that neither constitute substantial parts of the database nor encompass the structure of this database.
+    The remaining definition of this field is the same as for :field:`available_licenses`.
```
Comment (vaitkus) [inline optimade.rst:1270]: Maybe we should specify if the value of `available_licenses_for_entries` is unknown, the licenses from the `available_licenses` apply?

Comment (merkys) [reply optimade.rst:1270]: Right, we should state how these two fields interact.

File: optimade.rst [hunk 5]
```diff
@@ -1258,12 +1258,16 @@ The single resource object's response dictionary MUST include the following fiel
...
+    If the licensing information provided via the field :field:`license` omits licensing options specified in :field:`available_licenses`, or if it otherwise contradicts them, a client MUST still be allowed to interpret the inclusion of a license in :field:`available_licenses` as a full commitment from the database that the content, metadata and database structure are available, without exceptions, under the respective licenses.
+    If the database cannot make that commitment, e.g., if only part of the content, metadata or database structure is available under a license, the corresponding license identifier MUST NOT appear in :field:`available_licenses` (but, rather, the field :field:`license` is to be used to clarify the licensing situation.)
     An empty list indicates that none of the SPDX licenses apply for the entirety of the database and that the licensing situation is clarified in human readable form in the field :field:`license`.
+    An unknown value means that the database makes no commitment to its content, metadata and/or database structure data being covered by any of the SPDX licenses.
```
Comment (ml-evs) [inline optimade.rst:1267]:
```suggestion
    An unknown value (i.e., :val:`null` or missing) means that the database makes no commitment to its content, metadata and/or database structure data being covered by any of the SPDX licenses.
```

File: optimade.rst (hunk 4 above)
Comment (ml-evs) [reply optimade.rst:1270]: I think we need an example of the two scenarios, as this is quite abstract to me

File: optimade.rst (hunk 5 above)
Comment (merkys) [reply optimade.rst:1267]: I do not think this is necessary, as the specification already explains what "unknown value" means (it is a term in the specification), see "Properties with an unknown value" or "JSON Response Schema: Common Fields" sections for reference.

File: optimade.rst (hunk 4 above)
Comment (merkys) [reply optimade.rst:1270]: @ml-evs let's say a whole database is licensed as `CC0-1.0`, and it chooses to output just `available_licenses` field with `available_licenses_for_entries` set to unknown, as `available_licenses_for_entries` only pertains to a subset of the database (entry data). As I understand, @vaitkus asked to clarify the text to mean "look at `available_licenses` if `available_licenses_for_entries` is missing". As @rartino nicely put yesterday, `available_licenses_for_entries` is just an additional list of licenses for entry data, on top of `available_licenses`.

Now consider the case when `available_licenses` is undefined, but `available_licenses_for_entries` contains `CC0-1.0`. This should mean that the entry data is licensed under `CC0-1.0`, but there should be no assumptions about the remaining copyrightable content of the database, apart from entry data. Does this make sense?

**Review by rartino — CHANGES_REQUESTED** [review]
I have a few proposals that I think helps improve the clarity and stay clear of things possible to be interpreted in unintentional ways.

File: optimade.rst [hunk 6]
```diff
@@ -1258,12 +1258,17 @@ The single resource object's response dictionary MUST include the following fiel
...
+    If the database cannot make that commitment, e.g., if only part of the database is available under a license, the corresponding license identifier MUST NOT appear in :field:`available_licenses` (but, rather, the field :field:`license` is to be used to clarify the licensing situation.)
+    An empty list indicates that none of the SPDX licenses apply and that the licensing situation is clarified in human readable form in the field :field:`license`.
+    An unknown value means that the database makes no commitment.
+
+  - **available\_licenses\_for\_entries**: List of `SPDX license identifiers <https://spdx.org/licenses/>`__ specifying a set of additional alternative licenses under which the client is granted access to individual database entries, their metadata and data extracts that do not constitute substantial parts of the database.
+    If :field:`available_licenses` is known, the licenses provided in :field:`available_licenses_for_entries` are additional alternative licenses granting access to individual database entries, their metadata and data extracts that do not constitute substantial parts of the database.
```
[the remark this answers is not in the document]
Comment (merkys) [reply optimade.rst:1270]: Indeed, @vaitkus pointed this out to me too.

File: optimade.rst [hunk 7]
```diff
@@ -1258,12 +1258,17 @@ The single resource object's response dictionary MUST include the following fiel
...
-    If the database cannot make that commitment, e.g., if only part of the data is available under a license, the corresponding license identifier MUST NOT appear in :field:`available_licenses` (but, rather, the field :field:`license` is to be used to clarify the licensing situation.)
-    An empty list indicates that none of the SPDX licenses apply for the entirety of the database and that the licensing situation is clarified in human readable form in the field :field:`license`.
+  - **available\_licenses**: List of `SPDX license identifiers <https://spdx.org/licenses/>`__ specifying a set of alternative licenses under which the client is granted access to all the individual entries, all metadata, and the content and structure of the database itself.
+    If the database is available under multiple alternative licenses, identifiers of these multiple licenses SHOULD be provided to let clients know under which conditions they can be used.
```
[the remark this answers is not in the document]
Comment (merkys) [reply optimade.rst:1262]: Good point, I struggled with finding a better wording here.

File: optimade.rst [hunk 8]
```diff
@@ -1258,12 +1258,17 @@ The single resource object's response dictionary MUST include the following fiel
...
-    If the licensing information provided via the field :field:`license` omits licensing options specified in :field:`available_licenses`, or if it otherwise contradicts them, a client MUST still be allowed to interpret the inclusion of a license in :field:`available_licenses` as a full commitment from the database that the data and metadata is available, without exceptions, under the respective licenses.
-    If the database cannot make that commitment, e.g., if only part of the data is available under a license, the corresponding license identifier MUST NOT appear in :field:`available_licenses` (but, rather, the field :field:`license` is to be used to clarify the licensing situation.)
-    An empty list indicates that none of the SPDX licenses apply for the entirety of the database and that the licensing situation is clarified in human readable form in the field :field:`license`.
+  - **available\_licenses**: List of `SPDX license identifiers <https://spdx.org/licenses/>`__ specifying a set of alternative licenses under which the client is granted access to all the individual entries, all metadata, and the content and structure of the database itself.
```
[the remark this answers is not in the document]
Comment (merkys) [reply optimade.rst]: I think this is fine, but I leave this unresolved as a reminder for myself to look into how other people describe similar cases (for example, how Debian describes software it distributes).

File: optimade.rst (hunk 4 above)
Comment (rartino) [reply optimade.rst]: > Now consider the case when `available_licenses` is undefined, but `available_licenses_for_entries` contains `CC0-1.0`. This should mean that the entry data is licensed under `CC0-1.0`, but there should be no assumptions about the remaining copyrightable content of the database, apart from entry data. Does this make sense?

This is precisely how I would interpret it (and what I want it to mean).

File: optimade.rst (hunk 8 above)
[the remark this answers is not in the document]
Comment (merkys) [reply optimade.rst]: I looked at FAIRsharing (for example [FAIRsharing page for the COD](https://fairsharing.org/FAIRsharing.7mm5g5)), which lists databases and their licenses (among other properties), and it has the following header above each indicated database license:

>  This is a subset of licences recommended by the resource or most commonly used by its users.

This formulation seems to be written on the safer side, but FAIRsharing provides license links for informational use only while OPTIMADE aims to make them macine-readable and -actionable.

**Review by rartino — APPROVED** [review]
I guess I was holding off for other conversions to settle; but I think the current version is good enough to merge. Thanks for the work on this!

**Comment by merkys** [comment]
Thanks @rartino. It would be nice to get approvals from more OPTIMADE adopters as this PR changes the interpretation of license properties (new in this release, though).

**Comment by rartino** [comment]
@ml-evs In the spirit of cutting down on our PR log - since you've commented on this previously - do you still think there are remaining issues, or could you consider approving it?

@merkys 
> It would be nice to get approvals from more OPTIMADE adopters as this PR changes the interpretation of license properties (new in this release, though). 

This PR clarifies a few situations that were previously unclear, and extends what can be expressed. But as far as I see, no behavior that was well-defined with the previous PR *changes* by this PR. (Feel free to enlighten me if I'm missing something.)

**Review by ml-evs — APPROVED** [review]
I think the current wording is clear to me (and circumvents my need for examples). Thanks all!

**Comment by rartino** [comment]
@merkys Are you comfortable 'resolving' the two remaining conversations and merging this? (Apparently leaving conversations unresolved is now a blocker that cannot be overridden - I'm not sure how I feel about that.)

**Comment by merkys** [comment]
@rartino
> 
> @merkys
> 
> > It would be nice to get approvals from more OPTIMADE adopters as this PR changes the interpretation of license properties (new in this release, though).
> 
> This PR clarifies a few situations that were previously unclear, and extends what can be expressed. But as far as I see, no behavior that was well-defined with the previous PR _changes_ by this PR. (Feel free to enlighten me if I'm missing something.)

This is correct. I guess I tried to be extra-safe here with talking about the changes to the interpretation.

I have just marked all the remaining conversations about the text as resolved.

**Pull request merged**